llrt_utils/src/bytes.rs

```diff
@@ -1,8 +1,247 @@
 // Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 // SPDX-License-Identifier: Apache-2.0
+
+use std::borrow::Cow;
+
 use rquickjs::{ArrayBuffer, Coerced, Ctx, Exception, IntoJs, Object, Result, TypedArray, Value};
 
-use super::result::ResultExt;
+#[derive(Clone)]
+pub enum ObjectBytes<'js> {
+    U8Array(TypedArray<'js, u8>),
+    I8Array(TypedArray<'js, i8>),
+    U16Array(TypedArray<'js, u16>),
+    I16Array(TypedArray<'js, i16>),
+    U32Array(TypedArray<'js, u32>),
+    I32Array(TypedArray<'js, i32>),
+    U64Array(TypedArray<'js, u64>),
+    I64Array(TypedArray<'js, i64>),
+    F32Array(TypedArray<'js, f32>),
+    F64Array(TypedArray<'js, f64>),
+    DataView(ArrayBuffer<'js>),
+    Vec(Option<Vec<u8>>),
+}
+
+impl<'js> ObjectBytes<'js> {
+    const DETACHED_ERROR: &'static str = "ArrayBuffer is detached";
+
+    pub fn from(ctx: &Ctx<'js>, value: Value<'js>) -> Result<Self> {
+        Self::from_offset(ctx, value, 0, None)
+    }
+
+    pub fn from_offset(
+        ctx: &Ctx<'js>,
+        value: Value<'js>,
+        offset: usize,
+        length: Option<usize>,
+    ) -> Result<Self> {
+        if value.is_undefined() {
+            return Ok(ObjectBytes::Vec(Some(vec![])));
+        }
+        if let Some(bytes) = get_string_bytes(&value, offset, length)? {
+            return Ok(ObjectBytes::Vec(Some(bytes)));
+        }
+        if let Some(bytes) = get_array_bytes(&value, offset, length)? {
+            return Ok(ObjectBytes::Vec(Some(bytes)));
+        }
+
+        if let Some(obj) = value.as_object() {
+            if let Some(bytes) = Self::from_array_buffer(obj)? {
+                return Ok(bytes);
+            }
+        }
+
+        if let Some(bytes) = get_coerced_string_bytes(&value, offset, length) {
+            return Ok(ObjectBytes::Vec(Some(bytes)));
+        }
+
+        Err(Exception::throw_message(
+        ctx,
+        "value must be typed DataView, Buffer, ArrayBuffer, Uint8Array or interpretable as string",
+    ))
+    }
+
+    pub fn get_bytes(&mut self) -> Cow<'_, [u8]> {
+        let cow = match self {
+            ObjectBytes::U8Array(array) => {
+                Cow::Borrowed(array.as_bytes().expect(Self::DETACHED_ERROR))
```

Comment:
I would really prefer to return a `Result` since in my context if a user plugin script is screwing up, I don't want to crash my whole process. Though the buffer should not be detached since we hold a reference.

Reply:
I was considering this myself and you have a valid point. We should not be able to detach a buffer unless we do some really strange stuff. Also there are methods in `rquickjs` that also panics if buffer is detected so leaving this as is for now 🙂